packages/controllers/src/permissions/Permission.ts

```diff
@@ -0,0 +1,476 @@
+> = (args: RestrictedMethodOptions<Params>) => Promise<Result>;
+
+/**
+ * A synchronous or asynchronous restricted method implementation.
+ *
+ * @template Params - The JSON-RPC parameters of the restricted method.
+ * @template Result - The JSON-RPC result of the restricted method.
+ */
+export type RestrictedMethod<
+  Params extends RestrictedMethodParameters,
+  Result extends Json,
+> =
+  | SyncRestrictedMethod<Params, Result>
+  | AsyncRestrictedMethod<Params, Result>;
+
+export type GenericRestrictedMethod = RestrictedMethod<
+  RestrictedMethodParameters,
+  Json
+>;
+
+export type ValidateRestrictedMethod<
+  MethodImplementation extends RestrictedMethod<any, any>,
+> = MethodImplementation extends (args: infer Options) => Json | Promise<Json>
+  ? Options extends RestrictedMethodOptions<RestrictedMethodParameters>
+    ? MethodImplementation
+    : never
+  : never;
+
+export type PermissionFactory<
+  TargetPermission extends GenericPermission,
+  RequestData extends Record<string, unknown>,
+> = (
+  options: PermissionOptions<TargetPermission>,
+  requestData?: RequestData,
+) => TargetPermission;
+
+export type PermissionValidator<TargetPermission extends GenericPermission> = (
+  permission: GenericPermission,
+  origin?: OriginString,
+  target?: TargetPermission['parentCapability'],
+) => void;
+
+/**
+ * A utility type for ensuring that the given permission target key conforms to
+ * our naming conventions.
+ *
+ * See the documentation for the distinction between target names and keys.
+ *
+ * @template Key - The target key string to apply the constraint to.
+ */
+type ValidateTargetKey<Key extends string> = Key extends `${string}_*`
+  ? Key
+  : Key extends `${string}_`
+  ? never
+  : Key extends `${string}*`
+  ? never
+  : Key;
+
+/**
+ * A type extended by any function.
+ */
+type GenericFunction = (...args: any[]) => any;
+
+export type PermissionSpecificationConstraint = {
+  /**
+   * The target resource of the permission. In other words, at the time of
+   * writing, the RPC method name.
+   */
+  targetKey: string;
+
+  /**
+   * An array of the caveat types that may be added to instances of this
+   * permission.
+   */
+  allowedCaveats?: Readonly<NonEmptyArray<string>>;
+
+  /**
+   * The implementation of the restricted method that the permission
+   * corresponds to.
+   */
+  methodImplementation: GenericFunction;
+
+  /**
+   * The factory function used to get permission objects. Permissions returned
+   * by this function are presumed to valid, and they will not be passed to the
+   * validator function associated with this specification (if any). In other
+   * words, the factory function should validate the permissions it creates.
+   *
+   * If no factory is specified, the {@link Permission} constructor will be
+   * used, and the validator function (if specified) will be called on newly
+   * constructed permissions.
+   */
+  factory?: PermissionFactory<any, Record<string, unknown>>;
```

Comment:
These generic parameters don't seem useful. They're never set to anything but these values.

I wish there was a way to ensure the permission it created had a `parentCapability` that matched `targetKey` :thinking:. I can't see a way though.

Reply:
Oh, we can do it with a conditional type that wraps this one. Just not within this same type declaration.

Reply:
Oh, you've already done that with `ValidatePermissionSpecification`! Nice.

Reply:
Oh, no, just `methodImplementation` is narrowed there. Not `factory` and `validator`. Well, there's an opportunity here to ensure the factory and validator match the expected target key, if they're defined.